### PR TITLE
Move static Transcript methods and attributes into a Util class.

### PR DESCRIPTION
**Pull request by amuamushu**
Doing so moves the static method and attributes in a single spot, separate from the rest of the logic for TranscriptServlet.

### CODE REVIEW

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.w3c.dom.Node;
+
+/** Provides methods to create TranscriptLine Entities and TranscriptLine objects. */
+public class TranscriptLineUtil {
+  public static final String XML_URL_TEMPLATE = "http://video.google.com/timedtext?lang=en&v=";
+  public static final String ATTR_START = "start";
+  public static final String ATTR_DURATION = "dur";
+  public static final String TAG_TEXT = "text";
+
+  public static final String PARAM_LECTURE = "lecture";
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:33]: I don't think request parameters belong here

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  public static final String PARAM_LECTURE_ID = "id";
+  public static final String PARAM_VIDEO_ID = "video";
+  public static final String ENTITY_KIND = "TranscriptLine";
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:36]: change to `KIND`

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  public static final String PROP_LECTURE = "lecture";
+  public static final String PROP_START = "start";
+  public static final String PROP_DURATION = "duration";
+  public static final String PROP_CONTENT = "content";
+  public static final String PROP_END = "end";
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:42]: can you remove `PROP_` from these?

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.concurrent.TimeUnit;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+
+/** Provides methods to create TranscriptLine Entities and TranscriptLine objects. */
+public class TranscriptLineUtil {
+  public static final String XML_URL_TEMPLATE = "http://video.google.com/timedtext?lang=en&v=";
+  public static final String ATTR_START = "start";
+  public static final String ATTR_DURATION = "dur";
+  public static final String TAG_TEXT = "text";
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:31]: This is all unrelated to the transcriptline autovalue/entities.  I think it makes more sense for them to be in the servlet where you're actually handling the XML

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java (hunk 3 above)
Comment (jaysaleh) [reply src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:42]: ^^

File: src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import org.mockito.junit.MockitoRule;
+import org.w3c.dom.Element;
+
+@RunWith(JUnit4.class)
+public final class TranscriptLineUtilTest {
+  @Rule public final MockitoRule mockito = MockitoJUnit.rule();
+  @Mock private Element node;
+  
+  private LocalDatastoreServiceTestConfig datastoreConfig =
+      (new LocalDatastoreServiceTestConfig()).setNoStorage(true);
```
Comment (jaysaleh) [inline src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java:46]: Can this be moved into the localServiceHelper?

Comment (amuamushu) [reply src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java]: Moved!

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        .setDuration(duration)
+        .setEnd(end)
+        .setContent(content)
+        .build();
+  }
+
+  /**
+   * Creates a line entity using the attributes from {@code node} and {@code lectureId}.
+   */
+  // TODO: Update createEntity to take in the attributes rather than the nodes.
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:66]: Ah didn't see this TODO.  Disregard suggestion above

Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:66]: The one about moving parsing into servlet

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java [hunk 7]
```diff
@@ -251,11 +251,10 @@ private int countEntitiesInDatastore(String lectureKeyId) {
     return datastore.prepare(filteredQuery(lectureKeyId)).countEntities(withLimit(100));
   }
 
-  private Query filteredQuery(String lectureKeyId) {
-    Key lectureKeyKey =
-        KeyFactory.createKey(TranscriptServlet.PARAM_LECTURE, Long.parseLong(lectureKeyId));
-    Filter lectureKeyFilter =
-        new FilterPredicate(TranscriptLine.PROP_LECTURE, FilterOperator.EQUAL, lectureKeyKey);
-    return new Query(TranscriptLine.ENTITY_KIND).setFilter(lectureKeyFilter);
+  private Query filteredQuery(String lectureId) {
```
Comment (n-wach) [inline src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java:254]: I notice you're passing the ID around as a string and calling parseLong within each method.  It might be better to parse the ID once and pass the long (or resulting Key) around instead.  This applies outside of tests too.

File: src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+  @After
+  public void tearDown() {
+    localServiceHelper.tearDown();
+  }
+
+  @Test
+  public void fromEntity_transcriptLineSuccessfullyCreated() throws IOException {
+    Date testDate = new Date();
+    Key testLectureKey = KeyFactory.createKey(TranscriptServlet.PARAM_LECTURE, "Test Key Id");
```
Comment (n-wach) [inline src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java:63]: I thought IDs were integers?

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        .setLectureKey(lectureKey)
+        .setStart(start)
+        .setDuration(duration)
+        .setEnd(end)
+        .setContent(content)
+        .build();
+  }
+
+  /**
+   * Creates a line entity using the attributes from {@code node} and {@code lectureId}.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:64]:
```suggestion
   * Creates a transcript line entity using the attributes from {@code node} and {@code lectureId}.
```

**Comment by amuamushu** [comment]
This Pull Request resolves issue #51

File: src/main/java/com/googleinterns/zoomtube/data/TranscriptLine.java [hunk 10]
```diff
@@ -24,46 +23,37 @@
...
-   * @param content The text content of the transcript line.
-   */
-  public static TranscriptLine create(
-      Key transcriptKey, Key lectureKey, Date start, Date duration, Date end, String content) {
-    return new AutoValue_TranscriptLine(transcriptKey, lectureKey, start, duration, end, content);
-  }
-
+  /** The key for the transcript. */
   public abstract Key transcriptKey();
+  /** The key for the lecture. */
```
Comment (jaysaleh) [inline src/main/java/com/googleinterns/zoomtube/data/TranscriptLine.java:28]: Maybe add a new line after each of these

Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/data/TranscriptLine.java:29]: done!

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.googleinterns.zoomtube.data.TranscriptLine;
+import com.googleinterns.zoomtube.servlets.TranscriptServlet;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+/** Provides methods to create TranscriptLine Entities and TranscriptLine objects. */
+public class TranscriptLineUtil {
+  // TODO: Move these constants back to TranscriptServlet because
```
[the remark this answers is not in the document]
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java]: done! Thanks for catching that!

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        .setDuration(duration)
+        .setEnd(end)
+        .setContent(content)
+        .build();
+  }
+
+  /**
+   * Creates a transcript line entity using the attributes from {@code node} and {@code lectureId}.
+   */
+  // TODO: Update createEntity to take in the attributes rather than the nodes.
```
[the remark this answers is not in the document]
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java]: done! Thanks for catching that!

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.googleinterns.zoomtube.data.TranscriptLine;
+import com.googleinterns.zoomtube.servlets.TranscriptServlet;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+/** Provides methods to create TranscriptLine Entities and TranscriptLine objects. */
+public class TranscriptLineUtil {
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:26]: Please make this a final class and add a private constructor to indicate this is a util class. Making final prevents inheritance and private constructor allows it not be instantiated.

Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java]: Gotcha. Thanks for letting me know!

File: src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+/** Provides methods to create TranscriptLine Entities and TranscriptLine objects. */
+public final class TranscriptLineUtil {
+  public static final String KIND = "TranscriptLine";
+  public static final String LECTURE = "lecture";
+  public static final String START = "start";
+  public static final String DURATION = "duration";
+  public static final String CONTENT = "content";
+  public static final String END = "end";
+
+  private TranscriptLineUtil() {}
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/utils/TranscriptLineUtil.java:33]: Please move this at then end because it is a useless method. Please see: The order you choose for the members and initializers of your class can have a great effect on learnability. However, there's no single correct recipe for how to do it; different classes may order their contents in different ways.

https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java [hunk 15]
```diff
@@ -69,9 +70,12 @@
       (new LocalDatastoreServiceTestConfig()).setNoStorage(true);
   private static final LocalServiceTestHelper localServiceHelper =
       new LocalServiceTestHelper(datastoreConfig);
-  private static final String LECTURE_ID_A = "123";
-  private static final String LECTURE_ID_B = "345";
-  private static final String LECTURE_ID_C = "234";
+  private static final String LECTURE_ID_A_AS_STRING = "123";
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java:73]: I would not keep both. You can create the long and then convert it to a string.

Comment (billie-cat) [reply src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java:73]: Look at https://docs.oracle.com/javase/7/docs/api/java/lang/Long.html for converting a Long to a String.

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java [hunk 16]
```diff
@@ -130,8 +134,8 @@ public void tearDown() {
 
   @Test
   public void doGet_getDataInDatastoreForShortVideo() throws ServletException, IOException {
-    putTranscriptLinesInDatastore(shortVideoTranscriptLines, LECTURE_ID_A);
-    when(request.getParameter(LectureUtil.ID)).thenReturn(LECTURE_ID_A);
+    putTranscriptLinesInDatastore(shortVideoTranscriptLines, LECTURE_ID_A_AS_LONG);
+    when(request.getParameter(LectureUtil.ID)).thenReturn(LECTURE_ID_A_AS_STRING);
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java:138]: There is no way for getParameter to return a long?

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java [hunk 17]
```diff
@@ -235,33 +239,32 @@ public void doGet_twoLecturesInDatastore_returnsOneLecture()
...
+      Entity lineEntity = new Entity(TranscriptLineUtil.KIND);
+      lineEntity.setProperty(TranscriptLineUtil.LECTURE, lectureKey);
       // Set dummy values because AutoValue needs all the values to create a TranscriptLine object.
-      lineEntity.setProperty(TranscriptLine.PROP_START, /* start= */ new Date());
-      lineEntity.setProperty(TranscriptLine.PROP_DURATION, /* duration= */ new Date());
-      lineEntity.setProperty(TranscriptLine.PROP_CONTENT, /* content= */ "");
-      lineEntity.setProperty(TranscriptLine.PROP_END, /* end= */ new Date());
+      lineEntity.setProperty(TranscriptLineUtil.START, /* start= */ new Date());
+      lineEntity.setProperty(TranscriptLineUtil.DURATION, /* duration= */ new Date());
+      lineEntity.setProperty(TranscriptLineUtil.END, /* end= */ new Date());
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java:250]: For this could you can use your util class?

File: src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    Float durationAsFloat = 23.32F;
+    Float endDateAsFloat = startDateAsFloat.floatValue() + durationAsFloat.floatValue();
+    long startDateAsLong = (long) 23.32;
+    long durationAsLong = (long) 23.32;
+    long lectureId = 1;
+
+    Entity actualEntity = TranscriptLineUtil.createEntity(
+        lectureId, TEST_CONTENT, startDateAsFloat, durationAsFloat, endDateAsFloat);
+    Key actualKey = KeyFactory.createKey(LectureUtil.KIND, lectureId);
+    Date actualStart = new Date(TimeUnit.SECONDS.toMillis(startDateAsLong));
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java:90]: These are all part of checking the results so move them to the third section. Also don't forget to declare them only right before you need them.

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java (hunk 15 above)
Comment (amuamushu) [reply src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java]: Done! I kept the lecture ids as `Long`

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java [hunk 19]
```diff
@@ -193,21 +194,21 @@ public void doGet_returnsLectureForLongVideoFromDatastore() throws ServletExcept
...
     when(request.getParameter(LectureUtil.VIDEO_ID)).thenReturn(LONG_VIDEO_ID);
-    when(request.getParameter(LectureUtil.ID)).thenReturn(LECTURE_ID_C);
+    when(request.getParameter(LectureUtil.ID)).thenReturn(LECTURE_ID_C.toString());
 
     transcriptServlet.doPost(request, response);
 
-    int actualQueryCount = entitiesInDatastoreCount(LECTURE_ID_C);
-    int expectedQueryCount = (transcriptLines(LONG_VIDEO_JSON)).size();
-    assertThat(actualQueryCount).isEqualTo(expectedQueryCount);
+    int actualQuery = entitiesInDatastoreCount(LECTURE_ID_C);
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java:201]: Why did you rename to actualQuery. It seems to me actualQueryCount was better because it indicates its a count.

File: src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java (hunk 18 above)
Comment (billie-cat) [reply src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java:91]: When I said declare them in the right before they are needed I meant:

Key actualKey = KeyFactory.createKey(LectureUtil.KIND, LECTURE_ID_A);
assertThat(actualEntity.getProperty(TranscriptLineUtil.LECTURE)).isEqualTo(actualKey);
assertThat(actualEntity.getProperty(TranscriptLineUtil.CONTENT)).isEqualTo(TEST_CONTENT);
Date actualStart = new Date(TimeUnit.SECONDS.toMillis(startDateAsLong));
assertThat(actualEntity.getProperty(TranscriptLineUtil.START)).isEqualTo(actualStart);
...

Remember this style guide recommendation: https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations

Comment (amuamushu) [reply src/test/java/com/googleinterns/zoomtube/utils/TranscriptLineUtilTest.java]: Sorry I think I skimmed over the last sentence.

Done!

File: src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java (hunk 19 above)
Comment (amuamushu) [reply src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java]: I think at the time, I considered count a filler word because int implies that it is a number. Looking at it again, it makes sense to switch it back to `actualQueryCount` since it being an `int` doesn't guarantee it's a count.

Done!

Comment (billie-cat) [reply src/test/java/com/googleinterns/zoomtube/servlets/TranscriptServletTest.java]: I thought that might be the reason you removed it. Had you written actualQueryInt then I would have said int is filler, but count does describe what the number represents.
